contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,457 @@
+        // Convert ETH amount to BPT amount
+        bptExpected = bptExpected.divPrecisely(bptRate);
+    }
+
+    function _lpDepositAll() internal virtual;
+
+    function _lpWithdraw(uint256 numBPTTokens) internal virtual;
+
+    function _lpWithdrawAll() internal virtual;
+
+    /**
+     * @notice Balancer returns assets and rateProviders for corresponding assets ordered
+     * by numerical order.
+     */
+    function getPoolAssets() internal view returns (IERC20[] memory assets) {
+        (assets, , ) = balancerVault.getPoolTokens(balancerPoolId);
+    }
+
+    /**
+     * @dev If an asset is rebasing the Balancer pools have a wrapped versions of assets
+     * that the strategy supports. This function converts the pool(wrapped) asset
+     * and corresponding amount to strategy asset.
+     */
+    function toPoolAsset(address asset, uint256 amount)
+        internal
+        view
+        returns (address poolAsset, uint256 poolAmount)
+    {
+        poolAmount = 0;
+        if (asset == stETH) {
+            poolAsset = wstETH;
+            if (amount > 0) {
+                poolAmount = IWstETH(wstETH).getWstETHByStETH(amount);
+            }
+        } else if (asset == frxETH) {
+            poolAsset = sfrxETH;
+            if (amount > 0) {
+                poolAmount = IERC4626(sfrxETH).convertToShares(amount);
+            }
+        } else {
+            poolAsset = asset;
+            poolAmount = amount;
+        }
+    }
+
+    /**
+     * @dev Converts a Vault collateral asset to a Balancer pool asset.
+     * stETH becomes wstETH, frxETH becomes sfrxETH and everything else stays the same.
+     * @param asset Address of the Vault collateral asset.
+     * @return Address of the Balancer pool asset.
+     */
+    function toPoolAsset(address asset) internal view returns (address) {
+        if (asset == stETH) {
+            return wstETH;
+        } else if (asset == frxETH) {
+            return sfrxETH;
+        }
+        return asset;
+    }
+
+    /**
+     * @dev Converts rebasing asset to its wrapped counterpart.
+     */
+    function wrapPoolAsset(address asset, uint256 amount)
+        internal
+        returns (address wrappedAsset, uint256 wrappedAmount)
+    {
+        if (asset == stETH) {
+            wrappedAsset = wstETH;
+            if (amount > 0) {
+                wrappedAmount = IWstETH(wstETH).wrap(amount);
+            }
+        } else if (asset == frxETH) {
+            wrappedAsset = sfrxETH;
+            if (amount > 0) {
+                wrappedAmount = IERC4626(sfrxETH).deposit(
+                    amount,
+                    address(this)
+                );
+            }
+        } else {
+            wrappedAsset = asset;
+            wrappedAmount = amount;
+        }
+    }
+
+    /**
+     * @dev Converts wrapped asset to its rebasing counterpart.
+     */
+    function unwrapPoolAsset(address asset, uint256 amount)
+        internal
+        returns (uint256 wrappedAmount)
```

Comment:
Isn't this the _unwrapped_ amount, rather than the _wrapped_ amount?

Reply:
goo catch thanks, changed